theseus/theseus_layer.py

```diff
@@ -93,3 +103,88 @@ def device(self) -> torch.device:
     @property
     def dtype(self) -> torch.dtype:
         return self.objective.dtype
+
+
+def _forward(objective, optimizer, optimizer_kwargs, input_data):
+    objective.update(input_data)
+    info = optimizer.optimize(**optimizer_kwargs)
+    vars = [var.data for var in objective.optim_vars.values()]
+    return vars, info
+
+
+class TheseusLayerDLMForward(torch.autograd.Function):
```

Comment:
Overall looks good. Once the functionality is settled and unit tested we should think about compartmentalizing this maybe as separate files/modules outside of TheseusLayer.

Reply:
@bamos, @luisenp, Joe will likely add more things in this vein for implicit differentiation for belief propagation - GBP (which will branch off from NonlinearOptimizer). So we should think a more general API here if possible.

Reply:
Using autograd.Function also tends to introduce a bit of overhead and might not be very amenable to the theseus' object-oriented api. For one, it doesn't handle containers for tensors (e.g. dicts) very well.

Reply:
Any major limitation of this right now, @rtqichen, that we should address? If not we can flag an issue and come back to it after.

Reply:
None that I'm aware of feature-wise, but due to limitations of autograd.Function not being able to work with dicts, there's currently quite a bit of overhead from the use of python for-loops for converting between different containers for tensors (dicts, lists, subsets of lists). I can come back to this to see if I can speed up / remove some of this overhead later.